daft/ai/openai/text_embedder.py

```diff
@@ -0,0 +1,159 @@
+
+    from daft.ai.openai.typing import OpenAIProviderOptions
+    from daft.ai.typing import Embedding
+
+
+@dataclass(frozen=True)
+class _Profile:
+    """Model profiles contain various model-specific metadata.
+
+    Note:
+        This is a bit simpler than OO-inheritance to model the subtle
+        differences between the models. If there is a need for different
+        implementations, then it would make sense to
+    """
+
+    dimensions: EmbeddingDimensions
+
+
+_profiles: dict[str, _Profile] = {
+    "text-embedding-ada-002": _Profile(
+        dimensions=EmbeddingDimensions(size=1536, dtype=DataType.float32()),
+    ),
+    "text-embedding-3-small": _Profile(
+        dimensions=EmbeddingDimensions(size=1536, dtype=DataType.float32()),
+    ),
+    "text-embedding-3-large": _Profile(
+        dimensions=EmbeddingDimensions(size=3072, dtype=DataType.float32()),
+    ),
+}
+
+
+@dataclass
+class OpenAITextEmbedderDescriptor(TextEmbedderDescriptor):
+    provider_name: str
+    provider_options: OpenAIProviderOptions
+    model_name: str
+    model_options: Options
+
+    def __post_init__(self) -> None:
+        if self.model_name not in _profiles:
+            supported_models = ", ".join(_profiles.keys())
+            raise ValueError(
+                f"Unsupported OpenAI embedding model '{self.model_name}', expected one of: {supported_models}"
+            )
+
+    def get_provider(self) -> str:
+        return "openai"
+
+    def get_model(self) -> str:
+        return self.model_name
+
+    def get_options(self) -> Options:
+        return self.model_options
+
+    def get_dimensions(self) -> EmbeddingDimensions:
+        return _profiles[self.model_name].dimensions
+
+    def instantiate(self) -> TextEmbedder:
+        return OpenAITextEmbedder(
+            client=OpenAI(**self.provider_options),
+            model=self.model_name,
+        )
+
+
+class OpenAITextEmbedder(TextEmbedder):
+    """The OpenAI TextEmbedder will batch across rows, and split a large row into a batch request when necessary.
+
+    Note:
+        This limits us to 300k tokens per row which is a reasonable start.
+        This implementation also uses len(text)*5 to estimate token count
+        which is conservative and O(1) rather than being perfect with tiktoken.
+    """
+
+    _client: OpenAI
+    _model: str
+
+    def __init__(self, client: OpenAI, model: str, zero_on_failure: bool = True):
+        self._client = client
+        self._model = model
+        self._zero_on_failure = zero_on_failure
+
+    def embed_text(self, text: list[str]) -> list[Embedding]:
+        embeddings: list[Embedding] = []
+        curr_batch: list[str] = []
+        curr_batch_token_count: int = 0
+
+        batch_token_limit = 300_000
+        approx_tokens_per_char = 5  # rounding up from "1 token ≈ 4 characters"
+        input_text_token_limit = 8192
+        input_text_chars_limit = 1638  # rounding down from 8192 / 5
```

Comment:
This configuration does not take into account any rate limiting, only request size limiting.

| model | tokens/min | request/min | tokens/day |
-- | -- | -- | --
text-embedding-3-large | 1,000,000 TPM | 3,000 RPM | 3,000,000 TPD
text-embedding-3-small | 1,000,000 TPM | 3,000 RPM | 3,000,000 TPD
text-embedding-ada-002 | 1,000,000 TPM | 3,000 RPM | 3,000,000 TPD